fixed snappy decoder bug:   `HadoopSnappyDecoder::buffer` is invalid when copy multi blocks input data at the same place

lgbo-ustc opened the pull request:
Changelog category (leave one):

- Bug Fix 


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

When the inner readbuffer's buffer size is too small, NEED_MORE_INPUT in `HadoopSnappyDecoder` will run multi times (>=3)for one compressed block. This makes the input data be copied into the wrong place in  `HadoopSnappyDecoder::buffer`.


> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

lgbo-ustc committed:
fixed bug: call need_more_input repeatly, overwrite the buffer

lgbo-ustc committed:
add some usage test

lgbo-ustc committed:
add a test case

vdimir commented:
I checked if we have tests for snappy, and found one that was disabled. Could you add commit from https://github.com/vdimir/ClickHouse/tree/snappy_01060_avro to enable it into this PR?

vdimir reviewed (approved):
Let's wait CI

Also I checked if we have tests for snappy, and found one that was disabled. Could you add commit from [https://github.com/vdimir/ClickHouse/tree/snappy_01060_avro](https://github.com/vdimir/ClickHouse/tree/snappy_01060_avro?rgh-link-date=2022-03-09T13%3A26%3A00Z) to enable it into this PR?

lgbo-ustc commented on src/IO/tests/gtest_hadoop_snappy_decoder.cpp:
```diff
@@ -0,0 +1,38 @@
+using namespace DB;
+TEST(HadoopSnappyDecoder, repeat_need_more_input)
+{
+    String snappy_base64_content = "AAAl6gAAB67qSzQyMDIxLTA2LTAxAXh4eGIEACQzMTkuNzQyNDMKnjEAHDQyLjgyMTcynjEAIDI5Ni4yODQwNqIxABgyNy43MjYzqpMAGDMuNTIyMzSiYgAcNjUuNDk1OTeiMQAYOTQuNTg1NaYxABg4OC40NzgyojEAHDMyMS4zOTE1osQAHDM0Ni4xNTI3qjEAGDEuMjA3MTWm9QAQMi41MjamYQAcMjIuNTEyNDieYQAcMzMwLjI5MTKiGgIcMzIzLjAzNDiiwwAcMzE1LjA1MDmiYgAcNDM1Ljc2ODaqxAAUMS45NDA5nvQACDAuMP4rAEorABwzMDMuMjAyNaYZARgwOC4xOTEypugAGDQ ...
```
Ok

lgbo-ustc committed:
update tests

1. fixed code style in src/IO/tests/gtest_hadoop_snappy_decoder.cpp
2. enable tests 01060_avro.sh

lgbo-ustc commented:
1. fixed code style in [gtest_hadoop_snappy_decoder.cpp](https://github.com/ClickHouse/ClickHouse/pull/35116/files/8dc8c87fd1cea08e7e9875dd7774b02b1f3eef63#diff-32a1a18cd4cf0198a6ccaefc0727d496c89c31c786b867490fe572322410f452)
2. enable snappy test in [01060_avro.sh](https://github.com/vdimir/ClickHouse/commit/d304fa31de433af6b7df4d9f7ca3d851eec61dba#diff-f709fcd9cef80809c0d7ccda3cb5ac33638802e2198e6ed5fe41c0aba96db973)

lgbo-ustc committed:
fixed code style
